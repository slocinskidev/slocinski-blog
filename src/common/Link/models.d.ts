export type LinkVariant = 'nav' | 'regular' | 'hamburger';

export interface LinkProps {
  children: string | React.ReactNode;
  url: string;
  variant?: LinkVariant;
  ariaLabel?: string;
  customClass?: string;
  activeClass?: string;
}
