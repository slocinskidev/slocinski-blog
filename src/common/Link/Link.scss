@import 'styles/common/component-set';

.link {
  font-size: 16px;
  text-decoration: none;

  &--nav {
    padding: 0 1rem;

    @include breakpoint(lg-up) {
      padding: 0 1.75rem;
    }
  }

  &--hamburger {
    padding: 24px 0;
    @include font-size(26);

    &:last-of-type {
      padding: 24px 0 48px;
    }
  }
}
