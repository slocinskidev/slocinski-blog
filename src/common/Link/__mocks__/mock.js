import { LINKS } from 'shared/constants/tests';

export const LinkPropsMock = {
  url: LINKS.HOME,
  variant: 'nav',
  customClass: 'test-class',
};

export const classNamesMock = [LinkPropsMock.customClass, 'link--nav'];
