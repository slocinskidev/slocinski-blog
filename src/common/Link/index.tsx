export { default } from './Link';
export * from './models.d';
