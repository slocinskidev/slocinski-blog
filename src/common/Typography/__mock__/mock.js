export const typographyClassesMock = [
  'typography--h1',
  'typography--gutter-bottom-1',
  'typography--gutter-top-1',
  'typography--gutter-left-1',
  'typography--gutter-right-1',
  'typography--line-height-big',
  'typography--bold',
  'typography--italic',
];

export const TypographyPropsMock = {
  variant: 'h1',
  customClass: 'testClass',
  align: 'center',
  gutterBottom: 1,
  gutterTop: 1,
  gutterLeft: 1,
  gutterRight: 1,
  lineHeight: 'big',
  bold: true,
  italic: true,
};

export const bodyVariantClassname = 'typography--body1';

export const TEST_TXT = 'Test Lorem Ipsum';
