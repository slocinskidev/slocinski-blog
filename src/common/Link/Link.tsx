import React, { FC } from 'react';
import { Link as InternalLink } from 'gatsby';
import classNames from 'classnames';

import { LinkProps } from './models';

import './Link.scss';

const Link: FC<LinkProps> = ({
  children,
  url,
  variant = 'regular',
  ariaLabel,
  customClass,
  activeClass,
  ...rest
}) => {
  const linkClassnames = classNames('link', customClass, {
    [`link--${variant}`]: variant,
  });

  const isInternal = /^\/(?!\/)/.test(url);

  const renderLink = isInternal ? (
    <InternalLink
      className={linkClassnames}
      activeClassName={activeClass}
      aria-label={ariaLabel}
      to={url}
      {...rest}
    >
      {children}
    </InternalLink>
  ) : (
    <a className={linkClassnames} aria-label={ariaLabel} href={url} {...rest}>
      {children}
    </a>
  );

  return renderLink;
};

export default Link;
