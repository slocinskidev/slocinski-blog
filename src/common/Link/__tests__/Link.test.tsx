import React from 'react';
import { Link as GatsbyLink } from 'gatsby';
import { render, screen, userEvent } from 'testUtils';

import { classNamesMock, LinkPropsMock } from '../__mocks__/mock';
import Link from '../Link';
import { LinkProps } from '../models';

const { url } = LinkPropsMock;

const defaultProps: LinkProps = {
  ...(LinkPropsMock as LinkProps),
};

test('the Link component renders correctly with required and optional props', () => {
  const { rerender } = render(<Link {...defaultProps}>TestLink</Link>);

  const linkNode = screen.getByRole('link');

  // the component has been rendered
  expect(linkNode).toBeInTheDocument();

  // generating a snapshot, not for testing purposes but only to see what changes in the HTML structure during CR if the developer updates the snapshot
  expect(linkNode).toMatchSnapshot();

  // after click on link correct page appears
  userEvent.click(linkNode);
  expect(GatsbyLink).toHaveBeenCalledWith(expect.objectContaining({ to: url }), expect.anything());

  // link node receives correct classnames if provided
  expect(linkNode).toHaveClass(...classNamesMock);

  // rerender without optional props
  rerender(
    <Link url={LinkPropsMock.url} variant="regular">
      TestLink
    </Link>
  );

  // link node should not receive classnames if not provided
  expect(linkNode).not.toHaveClass(...classNamesMock);
});
