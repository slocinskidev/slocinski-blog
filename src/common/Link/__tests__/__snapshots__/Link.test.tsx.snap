// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`the Link component renders correctly with required and optional props 1`] = `
<a
  class="link test-class link--nav"
  href="/"
>
  TestLink
</a>
`;
